rekallpy/rekall/interval_set.py

```diff
@@ -809,32 +810,63 @@ def coalesce(self,
         Returns:
             A new IntervalSet of intervals that are disjoint along ``axis`` and
             are at least ``epsilon`` apart.
+            
         """
-
-        def update_output(output, intrvl):
-            if len(output) == 0:
-                output.append(intrvl)
-            else:
-                merge_candidate = output[-1]
+        if (len(self._intrvls) == 0):
+            return self
+
+        new_coalesced_intrvls = []
+        current_intrvls = []
+
+        for intrvl in self._intrvls:
+            for cur in current_intrvls:
+                #adds any intervals that occured before the start of intrvl
+                if not Bounds.cast({
+			        axis[0] : 't1',
+			        axis[1] : 't2'
+		        })(or_pred(overlaps(),
+                    before(max_dist=epsilon)))(cur, intrvl):
+                        new_coalesced_intrvls.append(cur)
+            #re-update contents of current_intrvls ==> contains all intervals that overlap 
+            current_intrvls = [
+                cur
+                for cur in current_intrvls
                 if Bounds.cast({
                         axis[0]: 't1',
                         axis[1]: 't2'
                 })(or_pred(overlaps(),
-                           before(max_dist=epsilon)))(merge_candidate, intrvl):
-                    output[-1] = Interval(
-                        bounds_merge_op(merge_candidate['bounds'],
+                           before(max_dist=epsilon)))(cur, intrvl)
+            ]
+
+            matched_intrvl = None
+            loc = len(current_intrvls) - 1
+
+            if len(current_intrvls) == 0:
+                current_intrvls.append(intrvl.copy())
+                continue
+            
+            if predicate is None:
+                matched_intrvl = current_intrvls[-1]
+            else:
+                for index, cur in enumerate(current_intrvls):
+                    if predicate(cur, intrvl):
+                        matched_intrvl = cur
+                        loc = index
+
+            if matched_intrvl is None:
```

Comment:
Add a comment - if we didn't find a matching Interval, it means `intrvl` is the start of a new coalescing interval